File: 0039-combination-sum/solution.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> curr = new ArrayList<>();

        backtrack(candidates,target,0,curr,res);
        return res;
    }

    public void backtrack(int[] candidates, int target, int start, List<Integer> currentCombination, List<List<Integer>> result ){
        if(target == 0){
            result.add(new ArrayList<>(currentCombination));
            return;
        }

        if(target<0){
            return;
        }

        for(int i = start; i< candidates.length; i++){
            currentCombination.add(candidates[i]);
            backtrack(candidates, target-candidates[i], i, currentCombination, result);
            currentCombination.remove(currentCombination.size() - 1);
        }

    }
}

File: 0039-combination-sum/solution.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []

        def backtrack(i, cur_sum, path):
            if cur_sum == target:
                res.append(path.copy())
                return
            if cur_sum > target or i >= len(candidates):
                return

            path.append(candidates[i])
            backtrack(i, cur_sum + candidates[i], path)

            path.pop()
            backtrack(i+1, cur_sum, path)

        backtrack(0, 0, [])

        return res
